akka-projection-core/src/main/scala/akka/projection/scaladsl/SourceProvider.scala

```diff
@@ -4,11 +4,13 @@
 
 package akka.projection.scaladsl
 
+import scala.concurrent.Future
+
 import akka.stream.scaladsl.Source
 
 trait SourceProvider[Offset, Envelope] {
 
-  def source(offset: Option[Offset]): Source[Envelope, _]
+  def source(offset: () => Future[Option[Offset]]): Future[Source[Envelope, _]]
```

Comment:
> the return type could be `Source[Envelope, _]`, instead of a `Future`

I introduced the `Future[Source[..]]` return value as a consequence of Future-based "get external offsets" argument that's passed to Alpakka Kafka.   It would be possible to remove the `Future` here, but when offsets are not available we need to get "beginning offsets" from Kafka, which is also returned as a `Future`.  I'll need to introduce blocking in the `KafkaSourceProvider` if we want to remove `Future`.

Reply:
> missing corresponding change in javadsl.SourceProvider

:+1:

Reply:
Future[Source] can be turned into Source with `Source.futureSource` but I agree that it's easier for users (SourceProvider implementors) to understand Future[Offset] => Future[Source]